### DISCUSSION TITLE
Nested for loop not working

### DISCUSSION THREAD

**Discussion by AniketKumar2500** [CsqLang/Csq#278] [Ideas]
**Case 1: Found Weird Behaviour**
I have tested a small code:
```
l := 5
for i in 1->l:
 print i
```
It worked but for:
```
l:=5
for i in 1->l:
 for j in 1->i:
  print j
```
It didn't give any output....

**Case 2: Checking memory footprint**
```
for(int i__iter = 1;i__iter < id("l").intVal;i__iter++){
memory[memory.size()-1] = i__iter;

allocateVar("j",0);
print(id("i"));
for(int j__iter = 1;j__iter < id("i").intVal;j__iter++){
memory[memory.size()-1] = j__iter;

print(id("j"));
}
}
```
After looking at converted C++ code i found that, i_iter is not getting changed for every iteration..

**Comment by gekkowrld** [comment]
Any ideas on how to fix it?

**Comment by AniketKumar2500** [comment]
I think we should recreate the AST visiting layer of For loop to convert to C++ code...

**Comment by gekkowrld** [comment]
Can't we just write a compiler instead of interpreting to c++ like carbon does?

**Comment by AniketKumar2500** [comment]
> Can't we just write a compiler instead of interpreting to c++ like carbon does?

But it will fail our major objective to gain speed, currently it's converted into c++ in order to access our core libraries which are responsible for fast speed and dynamic typing.

**Comment by gekkowrld** [comment]
It can always be optimized, but if we convert to c++ then it will always be the limit.

**Comment by AniketKumar2500** [comment]
Hmm reasonable so we should implement it..

**Comment by gekkowrld** [comment]
[Suggestion/Proposal]
That would require complete rewrite of the project (and refinement of the syntax.
If it is going to be written in python #164  then this resources should be helpful:

- <https://github.com/windelbouwman/ppci>
- <https://www.reddit.com/r/Compilers/comments/17esxtt/how_do_you_write_a_compiler_in_python/>
- <https://mathspp.com/blog/building-a-python-compiler-and-interpreter#:~:text=Yes.,t%20compiled%20into%20machine%20code.>
- <https://medium.com/@marcelogdeandrade/writing-your-own-programming-language-and-compiler-with-python-a468970ae6df>

If it is going to be written using another language like C/C++ or others, there are plenty of materials out there.

The first step is to define the syntax (which is already done as of now, phew :) ) then what is important and so on.

I would suggest a completely different experimental repo to do this so as not to affect the current progress.
